app/controllers/carts_controller.rb

```diff
@@ -0,0 +1,92 @@
+class CartsController < ApplicationController
+  before_action :set_cart, only: [:show, :edit, :update, :destroy]
+
+  # GET /carts
+  # GET /carts.json
+  def index
+    @carts = Cart.all
+  end
+
+  # GET /carts/1
+  # GET /carts/1.json
+  def show
+    @total_num_items = total_num_items
+    @total_cost_items = total_cost_items
+  end
+
+  # GET /carts/new
+  def new
+    @cart = Cart.new
+  end
+
+  # GET /carts/1/edit
+  def edit
+  end
+
+  # POST /carts
+  # POST /carts.json
+  def create
+    @cart = Cart.new(cart_params)
+
+    respond_to do |format|
+      if @cart.save
+        format.html { redirect_to @cart, notice: 'Cart was successfully created.' }
+        format.json { render :show, status: :created, location: @cart }
+      else
+        format.html { render :new }
+        format.json { render json: @cart.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  # PATCH/PUT /carts/1
+  # PATCH/PUT /carts/1.json
+  def update
+    respond_to do |format|
+      if @cart.update(cart_params)
+        format.html { redirect_to @cart, notice: 'Cart was successfully updated.' }
+        format.json { render :show, status: :ok, location: @cart }
+      else
+        format.html { render :edit }
+        format.json { render json: @cart.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  # DELETE /carts/1
+  # DELETE /carts/1.json
+  def destroy
+    @cart.destroy
+    respond_to do |format|
+      format.html { redirect_to carts_url, notice: 'Cart was successfully destroyed.' }
+      format.json { head :no_content }
+    end
+  end
+
+  # get total number of items in the cart or 0
+  def total_num_items
+    if @cart.line_items
+      return @cart.line_items.count
+    end
+    0
```

Comment:
With rails the `line_items` association is created whether or not there are any line items. So this should be able to be simplified

```suggestion
    @cart.line_items.count
```

Outside Rails, if you were creating an instance of a Ruby class that had an attribute with other objects in it, then yes you would need to check if it exists first. Unless that attribute is initialized as an empty array by default of course, which is basically what rails is doing with a `ActiveRecord::Associations::CollectionProxy` object.

Reply:
Cool Jesse thank you for this. I added it in as I ran into an issue when testing and this seemed to solve it, however I think the issue in the tests eventually turned out to be something else, and I didn't remove this. I'll take it out and test :)